akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala

```diff
@@ -498,72 +485,93 @@ private[remote] class ArteryTransport(_system: ExtendedActorSystem, _provider: R
     association(remoteAddress).quarantine(reason = "", uid.map(_.toLong))
   }
 
-  def outbound(outboundContext: OutboundContext): Sink[Send, Future[Done]] = {
+  def outbound(outboundContext: OutboundContext, compression: LiteralCompression): Sink[Send, Future[Done]] = {
     Flow.fromGraph(killSwitch.flow[Send])
       .via(new OutboundHandshake(outboundContext, handshakeTimeout, handshakeRetryInterval, injectHandshakeInterval))
-      .via(encoder)
+      .via(encoder(compression))
       .toMat(new AeronSink(outboundChannel(outboundContext.remoteAddress), ordinaryStreamId, aeron, taskRunner,
         envelopePool, giveUpSendAfter))(Keep.right)
   }
 
   def outboundLarge(outboundContext: OutboundContext): Sink[Send, Future[Done]] = {
+    // TODO I don't think large messages care about compression; sanity check though
+    val compression = NoLiteralCompression
+
     Flow.fromGraph(killSwitch.flow[Send])
       .via(new OutboundHandshake(outboundContext, handshakeTimeout, handshakeRetryInterval, injectHandshakeInterval))
-      .via(createEncoder(largeEnvelopePool))
+      .via(createEncoder(largeEnvelopePool, compression))
       .toMat(new AeronSink(outboundChannel(outboundContext.remoteAddress), largeStreamId, aeron, taskRunner,
         envelopePool, giveUpSendAfter))(Keep.right)
   }
 
-  def outboundControl(outboundContext: OutboundContext): Sink[Send, (OutboundControlIngress, Future[Done])] = {
+  def outboundControl(outboundContext: OutboundContext, compression: LiteralCompression): Sink[Send, (OutboundControlIngress, Future[Done])] = {
     Flow.fromGraph(killSwitch.flow[Send])
       .via(new OutboundHandshake(outboundContext, handshakeTimeout, handshakeRetryInterval, injectHandshakeInterval))
       .via(new SystemMessageDelivery(outboundContext, systemMessageResendInterval, remoteSettings.SysMsgBufferSize))
       .viaMat(new OutboundControlJunction(outboundContext))(Keep.right)
-      .via(encoder)
+      .via(encoder(compression))
       .toMat(new AeronSink(outboundChannel(outboundContext.remoteAddress), controlStreamId, aeron, taskRunner,
         envelopePool, Duration.Inf))(Keep.both)
 
     // FIXME we can also add scrubbing stage that would collapse sys msg acks/nacks and remove duplicate Quarantine messages
   }
 
-  def createEncoder(pool: EnvelopeBufferPool): Flow[Send, EnvelopeBuffer, NotUsed] =
-    Flow.fromGraph(new Encoder(localAddress, system, compression, pool))
+  // FIXME: Compression table must be owned by each channel instead of having a global one
+  private def createInboundCompressionTable(): LiteralCompression = {
+    val advertActorRef = (ref: String, id: Int) ⇒ {
+      val addr = ActorPath.fromString(ref).address // TODO expensive      
+      log.error(s"advertise compression to ${addr}, of [$ref] => [$id] ")
+      sendControl(addr, CompressionProtocol.ActorRefCompressionAdvertisement(addr, ref, id))
```

Comment:
How does the versioning work? Is that the `id`? I think it's also important to include the actor system `UID`, i.e. use `UniqueAddress`, so that we never ever use a table that doesn't match the incarnation